### DISCUSSION TITLE
Dynamic Completion Codes with CloudResearch & cognition.run

### DISCUSSION THREAD

**Discussion by csducate** [jspsych/jsPsych#1170] [Q&A] [answered]
Hi All,

I'm about to launch a study using cognition.run and CloudResearch. I saw a [previous discussion](https://github.com/jspsych/jsPsych/discussions/710) with code for how to add a dynamic completion code to a jsPsych study based on [this](https://www.cloudresearch.com/resources/blog/qualtrics-and-other-platforms-mechanical-turk-integration-with-cloudresearch/) CloudResearch article.

The original poster stated that they had to move the saveData function in order to make the code work. However, because I'm running on cognition.run, I'm not sure whether I need to do that since I'm not explicitly using the saveData function--cognition.run is handling data saving for me. 

If I don't, how would I save a record of the dynamic completion code? Or does CloudResearch handle that for me?

**Comment by jodeleeuw** [comment, marked as answer]
I haven't used CloudResearch, but my understanding is that the dynamic completion code is something that the participant can click on to complete the experiment. Is that right?

cognition.run saves the data as the experiment progresses, so you will get data even if participants exit the experiment before it is technically finished. The data files will probably be flagged as incomplete in cognition.run's storage.

You could probably stick the iframe code in the on_finish event for `jsPsych.init`. Something like this might work:

```js
jsPsych.init({
	timeline: timeline,
	on_finish: function(){
		jsPsych.getDisplayElement().innerHTML = '<iframe src="https://www.turkprime.com/TurkPrime/TakeLaunchedSurvey/DynamicKey" width="100%" height="200"></iframe>'
	}
}
```

**Reply by csducate** [reply]
The completion code is presented to the participant, who then copies and pastes it in the CloudResearch webpage to confirm. 

Your code works great! But it doesn't actually save the dynamic code in my data file. I may just have to run on CloudResearch and see how it works in practice and whether I even need to save the dynamic code independently of CloudResearch...
